File: Algorithms/src/heap/MaxHeap.java
package heap;

/**
 * Created by vasantbedadala on 2/15/16.
 */
public class MaxHeap {
    int[] heap;
    int N;
    int size;
    public MaxHeap(int size) {
        heap = new int[size + 1];
        this.N=0;
        this.size=size;
    }

    /**
     * Build heap from bottom up -- maxheapify
     * Leaves are good as they are 1 element heaps, so start with N/2
     * O(n)
     */
    public MaxHeap(int[] keys) {
        heap = new int[keys.length + 1];
        //start with index 1, to avoid null links
        for (int i = 0; i < keys.length; i++) {
            heap[i + 1] = keys[i];
            this.N++;
        }
        for (int i = N / 2; i >= 1; i--) {
            sink(i);
        }
    }

    public int max() {
       if(!isEmpty()) {
           return heap[1];
       }
        return -1;
    }

    public int delMax() {
        if(!isEmpty()) {
            int max = heap[1];
            exch(heap, 1, N--);
            sink(1);
            return max;
        }
        return -1;
    }

    private boolean isMaxHeap(int k) {
        int left = 2 * k;
        int right = 2 * k + 1;
        if (heap[k] <= left) return false;
        if (heap[k] <= right) return false;
        return isMaxHeap(left) && isMaxHeap(right);
    }

    private void sink(int key) {
        while (2 * key <=N) {
            int k = 2 * key;
            if (k<N && heap[k] < heap[k + 1]) {
                k = k + 1;
            }
            if (heap[key] > heap[k])
                break;
            exch(heap, key, k);
            key = k;
        }
    }


    private void swim(int key) {
        //while not at root and greater than parent
        while (key != 1 && heap[key] > heap[(int) Math.floor(key / 2)]) {
            exch(heap, key, (int) Math.floor(key / 2));
            key = (int) Math.floor(key / 2);
        }
    }

    private void exch(int[] x, int i, int j) {
        int temp = x[i];
        x[i] = x[j];
        x[j] = temp;
    }

    public void insert(int key) {
        this.N++;
        heap[N] = key;
        swim(N);
    }

    // helper function to double the size of the heap array
    private void resize(int capacity) {
        int[] temp = new int[capacity];
        for (int i = 1; i <= N; i++) {
            temp[i] = heap[i];
        }
        heap = temp;
    }

    public boolean isEmpty() {
        return this.N==0;
    }

    public int size() {
        return this.N;
    }
}

File: algorithms-python/heap/MaxHeap.py
import math
'''
Max heap implementaion based on
Algorithms by Sedgewick
'''

class MaxHeap:
    def __init__(self):
        self.pq=[]
        self.N=0;

    def __init__(self,keys):
        self.heapify(keys)

    def swim(self,n):
        while n!=1 and self.pq[n] > self.pq[math.floor(n/2)]:
            self.pq[n],self.pq[math.floor(n/2)] = self.pq[math.floor(n/2)],self.pq[n]
            n=math.floor(n/2)

    def sink(self,n):
        while 2*n < self.N:
            if self.pq[2*n] > self.pq[2*n+1]:
                k=n
            else:
                k=n+1
            if self.pq[n]<self.pq[k]:
                self.pq[n],self.pq[k]=self.pq[k],self.pq[n]
            else:
                continue


    def heapify(self,keys):
        self.N=0
        for n in range(keys.length):
            self.pq[n+1]=keys[n]
            self.N+=1

        for n in range(math.floor(keys.lenth/2),0):
            self.sink(n)

    def insert(self,n):
        self.N+=1
        self.pq[self.N]=n
        self.swim(n)

    def deleteMax(self,n):
        self.pq[1],self.pq[self.N] =self.pq[self.N],self.pq[1]
        self.N-=1;
        self.sink(n)

    def max(self):
        self.pq[1]







